Carbonblack Translate and Transmit Module

Collin-Richards opened the pull request:
Along with implementing the translate and transmit modules, there are a couple extra changes that were made:

- Added support for arrays of stix references. An example of it's usage can be seen in the opened_connection_refs.
- Created StartStopQualifier object and attached the start and stop timestamp to it. The existing qualifier object has a single string and the individual module was responsible for doing validation.
- Added module dialect support to the transmission module.

yurii-klymenko commented:
I see that status connector implementation is missing here: stix_shifter/stix_transmission/src/modules/carbonblack/carbonblack_connector.py

Collin-Richards commented:
I think that's intentional because the carbonblack api is synchronous. The other synchronous module, aws_security_hub, is also missing this. Also if people would like to test the code on a carbon black instance I can give you temporary access to an API key.

emishas commented on stix_shifter/stix_transmission/src/modules/carbonblack/carbonblack_connector.py:
```diff
@@ -0,0 +1,62 @@
+    def create_results_connection(self, search_id, offset, length):
+        try:
+            query = search_id
+            response = self.api_client.run_search(query)
```
is there a way to fetch particular data range? the offset/length parameters

Collin-Richards replied:
Yes. It supports a start and a rows parameter. I'll add that. That also made me realize that the api defaults to returning only 10 rows.

yurii-klymenko commented on stix_shifter/stix_transmission/src/modules/carbonblack/carbonblack_connector.py:
```diff
@@ -0,0 +1,62 @@
+            if len(response_json) > 0 and 200 <= response_code < 300:
+                return_obj['success'] = True
+            else:
+                return_obj['success'] = False
```
error answer is supposed to have an 'code' field. Please check other modules(qradar,bigfix and splunk) for error reporting. It may be needed to add error_mapper.py to implement proper error mapping.

Collin-Richards replied:
Ok. I'll add that as well.

Collin-Richards commented:
I've addressed the issues mentioned above.

After noticing that the number of results returned were defaulting to 10, I've done some testing with a different limit. Requesting 2000 records takes about 8 seconds with a simple query (process_name:cmd.exe) and an arbitrary offset. Note in the test environment this has around 72,000 total results. I've tried testing with sorting by a couple different fields in case certain fields were indexed in their database, but they all appear to have similar runtime.

delliott90 commented on stix_shifter/stix_translation/src/modules/carbonblack/README.md:
````diff
@@ -0,0 +1,377 @@
+## Example STIX Output Format
+
+```
+$ python3 main.py translate carbonblack results '{"id": "blah"}' '<copied results from previous query>'
````
You may want to indicate that '{"id": "blah"}' should be the identity object that is shown below as the first object in the 'objects' array.

delliott90 commented on stix_shifter/stix_translation/src/modules/carbonblack/carbonblack_data_mapping.py:
```diff
@@ -0,0 +1,32 @@
+
+class CarbonBlackDataMapper:
+    def __init__(self, options):
+        self.map_data = self._fetch_mappings()
```
I know in QRadar we first look for a mapping in `options['mapping']` in case the end-user want to override mappings on the data source. Do you think similar support is needed here? I realize by seeing this that I failed to add anything covering options in our documentation.

Collin-Richards replied:
I can add support for that back in. Due to the binary and process api having separate mappings it seemed weird to me a user overriding them.

delliott90 commented on stix_shifter/stix_translation/src/modules/carbonblack/carbonblack_query_constructor.py:
```diff
@@ -0,0 +1,252 @@
+        self.dmm = data_model_mapper
+        self.pattern = pattern
+        self.result_limit = result_limit
+        print(pattern)
```
Probably don't need these print statements.

delliott90 commented on stix_shifter/stix_translation/src/modules/carbonblack/carbonblack_query_constructor.py:
```diff
@@ -0,0 +1,252 @@
+            stripped = stripped.split('.', 1)[0]
+        return stripped
+
+    def _fomat_start_stop_qualifier(self, expression, dialect, qualifier : StartStopQualifier) -> str:
```
typo in "fomat"

delliott90 commented on stix_shifter/stix_translation/src/patterns/grammar/STIXPatternParser.py:
```diff
@@ -1596,6 +1596,12 @@ def TimestampLiteral(self, i:int=None):
         def STOP(self):
             return self.getToken(STIXPatternParser.STOP, 0)
 
+        def start_time(self):
```
I don't know if it's a good idea to add code directly to this file since it is generated from the STIXPattern.g4 file. The grammar files get generated by running 
`antlr4 -Dlanguage=Python3 -listener -visitor STIXPattern.g4`
Even though the g4 file is included in this project it's mostly a copy of https://github.com/oasis-open/cti-stix2-json-schemas/blob/master/pattern_grammar/STIXPattern.g4. I say mostly because it appears the g4 file used in the original Mitre translation project was not from that oasis repo. There are differences and a straight copy breaks a bunch of tests.

Collin-Richards replied:
Ah good catch, I didn't see that it was generated. I'll investigate this a bit further. I do think this should be done in a way so each parser doesn't need to duplicate a regex for the timestamp logic though.

delliott90 commented on stix_shifter/stix_transmission/stix_transmission.py:
```diff
@@ -16,11 +16,22 @@ class StixTransmission:
+        if module not in TRANSMISSION_MODULES :
+            raise NotImplementedError
+
+        dialect = None
```
If the dialect is specific to Carbon Black, it may be better to push this logic down to the Connector if possible. I guess I'm not 100% sure what the module looks like before it gets split.

delliott90 commented:
I intentionally called a transmission query with a bad query `'{"query": "username:SYSTEM", "dialect": "binary"}'`. When I called the results with the same query, I got the following error:
`{'success': False, 'error': 'Unhandled exception. Check logs for details.', 'code': 500}`. I wonder if we need some better error messaging here. If I force a transmission error with QRadar I get something like:
`{'success': False, 'error': 'Query f9faf4be-f9a7-4756-912c-a63e55c87ce1 does not exist', 'code': 901}
`. The QRadar module has an ErrorMapper class which takes care of this messaging.  I'm not sure what errors come back from the Carbon Black APIs.

Collin-Richards commented:
Yeah. We should at least be displaying the error from carbon black in that case. They aren't that rich, but it should be better then the current message.

delliott90 commented:
After speaking with Collin, we've decided to address the error messaging in a future PR.
